{
 "cells": [
  {
   "metadata": {},
   "cell_type": "raw",
   "source": [
    "\n",
    "import numpy as np\n",
    "import pandas as pd"
   ],
   "id": "d165debddb7f6e97"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-18T17:03:03.230373Z",
     "start_time": "2025-05-18T17:03:03.224199Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ],
   "id": "1e89b8c09154a849",
   "outputs": [],
   "execution_count": 2
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-18T17:16:39.312830Z",
     "start_time": "2025-05-18T17:16:39.306772Z"
    }
   },
   "cell_type": "code",
   "source": [
    "vcf_file = \"test_snp_postions.vcf\"\n",
    "vcf_file1 = \"GAsP_1163Samples.InDels.Feb2020.vcf.gz\"\n",
    "vcf_file2 = \"homo_sapiens-chrY.vcf.gz\"\n",
    "vcf_file3 = r\"D:\\genome_asia_main_file\\genome_asia_data\\Multiallelic\\GAsP_Multiallelic_Y_SNP_bgzip.vcf.gz\""
   ],
   "id": "a4cfd2efbb99138b",
   "outputs": [],
   "execution_count": 8
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-18T17:16:41.306597Z",
     "start_time": "2025-05-18T17:16:41.290762Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def get_vcf_header(vcf_file_input):\n",
    "    import gzip\n",
    "    header_line = \"\"\n",
    "    # set the opener if it is gzip\n",
    "    opener = gzip.open if vcf_file_input.endswith(\".gz\") else open\n",
    "    # open the file and find the header line\n",
    "    with opener(vcf_file_input, \"rt\") as file:\n",
    "        for line in file:\n",
    "            if line.startswith(\"#CHROM\"):\n",
    "                header_line = line\n",
    "                break\n",
    "            elif not line.startswith(\"#\"):\n",
    "                break\n",
    "    # format the header line to a list format\n",
    "    if header_line:\n",
    "        header_columns = header_line.strip(\"#\").strip(\"\\n\").split(\"\\t\")\n",
    "        return header_columns\n",
    "    else:\n",
    "        return \"Unable to find the header, may be not a vcf file(if csv proceed & ignore)\"\n",
    "\n",
    "fixed_columns = get_vcf_header(vcf_file3)  # change file name here >>\n",
    "print(f\"These are the columns found :\\n{fixed_columns}\")"
   ],
   "id": "da91857f91f14b05",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "These are the columns found :\n",
      "['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT', 'GA000431', 'GA000432', 'GA000433', 'GA000434', 'GA000435', 'GA000436', 'GA000437', 'GA000438', 'GA000439', 'GA000440', 'GA000441', 'GA000442', 'GA000443', 'GA000444', 'GA000445', 'GA000446', 'GA000447', 'GA000448', 'GA000449', 'GA000450', 'GA000451', 'GA000452', 'GA000453', 'GA000454', 'GA000455', 'GA000456', 'GA000457', 'GA000459', 'GA000460', 'GA000461', 'GA000462', 'GA000463', 'GA000464', 'GA000465', 'GA000466', 'GA000467', 'GA000468', 'GA000469', 'GA000470', 'GA000471', 'GA000472', 'GA000473', 'GA000474', 'GA000475', 'GA000476', 'GA000477', 'GA000478', 'GA000479', 'GA000480', 'GA000481', 'GA000482', 'GA000483', 'GA000484', 'GA000485', 'GA000486', 'GA000487', 'GA000488', 'GA000489', 'GA000490', 'GA000491', 'GA000492', 'GA000493', 'GA000494', 'GA000495', 'GA000496', 'GA000497', 'GA000498', 'GA000499', 'GA000518', 'GA000519', 'GA000521', 'GA000523', 'GA000524', 'GA000525', 'GA000526', 'GA000527', 'GA000528', 'GA000529', 'GA000530', 'GA000531', 'GA000532', 'GA000533', 'GA000534', 'GA000535', 'GA000536', 'GA000537', 'GA000538', 'GA000539', 'GA000540', 'GA000541', 'GA000542', 'GA000543', 'GA000544', 'GA000545', 'GA000546', 'GA000547', 'GA000548', 'GA000549', 'GA000550', 'GA000551', 'GA000552', 'GA000553', 'GA000554', 'GA000555', 'GA000556', 'GA000557', 'GA000558', 'GA000559', 'GA000560', 'GA000561', 'GA000562', 'GA000563', 'GA000564', 'GA000565', 'GA000566', 'GA000567', 'GA000568', 'GA000569', 'GA000570', 'GA000571', 'GA000572', 'GA000573', 'GA000632', 'GA000633', 'GA000634', 'GA000635', 'GA000636', 'GA000637', 'GA000638', 'GA000639', 'GA000640', 'GA000641', 'GA000642', 'GA000781', 'GA000782', 'GA000783', 'GA000784', 'GA000785', 'GA000786', 'GA000787', 'GA000788', 'GA000789', 'GA000790', 'GA000791', 'GA000792', 'GA000793', 'GA000794', 'GA000795', 'GA000796', 'GA000797', 'GA000798', 'GA000799', 'GA000800', 'GA000801', 'GA000802', 'GA000803', 'GA000804', 'GA000805', 'GA000806', 'GA000807', 'GA000808', 'GA000809', 'GA000810', 'GA000811', 'GA000812', 'GA000813', 'GA000814', 'GA000815', 'GA000816', 'GA000817', 'GA000818', 'GA000819', 'GA000820', 'GA000821', 'GA000822', 'GA000823', 'GA000824', 'GA000825', 'GA000826', 'GA000827', 'GA000828', 'GA000829', 'GA000830', 'GA000831', 'GA000832', 'GA000833', 'GA000834', 'GA000835', 'GA000836', 'GA000837', 'GA000838', 'GA000839', 'GA000840', 'GA000841', 'GA000842', 'GA000843', 'GA000844', 'GA000845', 'GA000846', 'GA000847', 'GA000848', 'GA000849', 'GA000850', 'GA000851', 'GA000852', 'GA000853', 'GA000854', 'GA000855', 'GA000856', 'GA000857', 'GA000858', 'GA000859', 'GA000860', 'GA000861', 'GA000862', 'GA000863', 'GA000864', 'GA000865', 'GA000866', 'GA000867', 'GA000870', 'GA000871', 'GA000872', 'GA000873', 'GA000874', 'GA000875', 'GA000876', 'GA000877', 'GA000878', 'GA000879', 'GA000880', 'GA000881', 'GA000882', 'GA000883', 'GA000884', 'GA000885', 'GA000886', 'GA000887', 'GA000888', 'GA000889', 'GA000890', 'GA000891', 'GA000892', 'GA000893', 'GA000896', 'GA000897', 'GA000899', 'GA000900', 'GA000901', 'GA000902', 'GA000903', 'GA000904', 'GA000905', 'GA000906', 'GA000907', 'GA000908', 'GA000909', 'GA000910', 'GA000912', 'GA000913', 'GA000914', 'GA000915', 'GA000916', 'GA000917', 'GA000918', 'GA000919', 'GA000920', 'GA000921', 'GA000922', 'GA000923', 'GA000924', 'GA000925', 'GA000926', 'GA000927', 'GA000928', 'GA000929', 'GA000930', 'GA000931', 'GA000932', 'GA000933', 'GA000934', 'GA000935', 'GA000936', 'GA000937', 'GA000938', 'GA000939', 'GA000940', 'GA000941', 'GA000942', 'GA000943', 'GA000944', 'GA000945', 'GA000946', 'GA000947', 'GA000948', 'GA000949', 'GA000950', 'GA000951', 'GA000952', 'GA000953', 'GA000954', 'GA000955', 'GA000956', 'GA000957', 'GA000958', 'GA000959', 'GA000960', 'GA000961', 'GA000963', 'GA000964', 'GA000965', 'GA000966', 'GA000967', 'GA000968', 'GA000969', 'GA000970', 'GA000971', 'GA000972', 'GA000973', 'GA000974', 'GA000975', 'GA000976', 'GA000977', 'GA000978', 'GA000979', 'GA000980', 'GA000981', 'GA000982', 'GA000983', 'GA000984', 'GA000985', 'GA000986', 'GA000987', 'GA000988', 'GA000989', 'GA000990', 'GA000991', 'GA000992', 'GA000993', 'GA000994', 'GA000995', 'GA000996', 'GA000997', 'GA000998', 'GA000999', 'GA001000', 'GA001001', 'GA001002', 'GA001003', 'GA001004', 'GA001005', 'GA001006', 'GA001007', 'GA001008', 'GA001009', 'GA001010', 'GA001011', 'GA001012', 'GA001013', 'GA001014', 'GA001016', 'GA001017', 'GA001018', 'GA001019', 'GA001020', 'GA001021', 'GA001022', 'GA001023', 'GA001024', 'GA001025', 'GA001026', 'GA001027', 'GA001028', 'GA001029', 'GA001030', 'GA001031', 'GA001032', 'GA001033', 'GA001034', 'GA001035', 'GA001037', 'GA001038', 'GA001039', 'GA001040', 'GA001041', 'GA001042', 'GA001043', 'GA001044', 'GA001045', 'GA001046', 'GA001047', 'GA001048', 'GA001049', 'GA001050', 'GA001051', 'GA001052', 'GA001053', 'GA001054', 'GA001055', 'GA001056', 'GA001057', 'GA001058', 'GA001059', 'GA001060', 'GA001061', 'GA001062', 'GA001063', 'GA001064', 'GA001065', 'GA001066', 'GA001067', 'GA001068', 'GA001069', 'GA001070', 'GA001071', 'GA001072', 'GA001073', 'GA001074', 'GA001075', 'GA001076', 'GA001077', 'GA001078', 'GA001079', 'GA001080', 'GA001081', 'GA001082', 'GA001083', 'GA001085', 'GA001086', 'GA001087', 'GA001088', 'GA001089', 'GA001090', 'GA001091', 'GA001092', 'GA001093', 'GA001094', 'GA001095', 'GA001096', 'GA001097', 'GA001098', 'GA001099', 'GA001100', 'GA001101', 'GA001102', 'GA001103', 'GA001104', 'GA001105', 'GA001106', 'GA001107', 'GA001108', 'GA001109', 'GA001110', 'GA001111', 'GA001112', 'GA001113', 'GA001114', 'GA001115', 'GA001116', 'GA001117', 'GA001118', 'GA001119', 'GA001120', 'GA001121', 'GA001122', 'GA001123', 'GA001124', 'GA001125', 'GA001126', 'GA001127', 'GA001128', 'GA001129', 'GA001130', 'GA001131', 'GA001132', 'GA001133', 'GA001134', 'GA001135', 'GA001136', 'GA001137', 'GA001138', 'GA001139', 'GA001140', 'GA001141', 'GA001142', 'GA001143', 'GA001144', 'GA001145', 'GA001146', 'GA001147', 'GA001148', 'GA001149', 'GA001150', 'GA001151', 'GA001152', 'GA001154', 'GA001155', 'GA001156', 'GA001157', 'GA001158', 'GA001159', 'GA001160', 'GA001161', 'GA001162', 'GA001163', 'GA001164', 'GA001165', 'GA001166', 'GA001167', 'GA001168', 'GA001169', 'GA001170', 'GA001171', 'GA001172', 'GA001173', 'GA001174', 'GA001175', 'GA001176', 'GA001177', 'GA001178', 'GA001179', 'GA001180', 'GA001181', 'GA001182', 'GA001183', 'GA001184', 'GA001186', 'GA001187', 'GA001188', 'GA001189', 'GA001190', 'GA001191', 'GA001192', 'GA001193', 'GA001195', 'GA001196', 'GA001197', 'GA001198', 'GA001199', 'GA001200', 'GA001201', 'GA001202', 'GA001203', 'GA001204', 'GA001205', 'GA001206', 'GA001207', 'GA001208', 'GA001209', 'GA001210', 'GA001211', 'GA001212', 'GA001214', 'GA001215', 'GA001217', 'GA001218', 'GA001219', 'GA001220', 'GA001221', 'GA001222', 'GA001223', 'GA001224', 'GA001225', 'GA001227', 'GA001228', 'GA001229', 'GA001230', 'GA001231', 'GA001232', 'GA001234', 'GA001240', 'GA001241', 'GA001242', 'GA001243', 'GA001244', 'GA001245', 'GA001246', 'GA001247', 'GA001248', 'GA001251', 'GA001252', 'GA001253', 'GA001254', 'GA001256', 'GA001257', 'GA001258', 'GA001259', 'GA001260', 'GA001262', 'GA001263', 'GA001264', 'GA001265', 'GA001266', 'GA001267', 'GA001268', 'GA001269', 'GA001271', 'GA001272', 'GA001273', 'GA001274', 'GA001275', 'GA001276', 'GA001277', 'GA001278', 'GA001279', 'GA001280', 'GA001281', 'GA001282', 'GA001284', 'GA001285', 'GA001286', 'GA001287', 'GA001288', 'GA001289', 'GA001292', 'GA001295', 'GA001296', 'GA001298', 'GA001299', 'GA001305', 'GA001307', 'GA001308', 'GA001309', 'GA001311', 'GA001312', 'GA001313', 'GA001314', 'GA001316', 'GA001319', 'GA001320', 'GA001321', 'GA001322', 'GA001323', 'GA001324', 'GA001325', 'GA001326', 'GA001327', 'GA001328', 'GA001329', 'GA001330', 'GA001331', 'GA001332', 'GA001333', 'GA001334', 'GA001335', 'GA001336', 'GA001337', 'GA001338', 'GA001339', 'GA001340', 'GA001341', 'GA001342', 'GA001343', 'GA001344', 'GA001345', 'GA001346', 'GA001347', 'GA001348', 'GA001349', 'GA001350', 'GA001351', 'GA001352', 'GA001354', 'GA001355', 'GA001356', 'GA001357', 'GA001361', 'GA001362', 'GA001364', 'GA001365', 'GA001366', 'GA001367', 'GA001368', 'GA001369', 'GA001370', 'GA001371', 'GA001372', 'GA001373', 'GA001374', 'GA001375', 'GA001376', 'GA001377', 'GA001378', 'GA001379', 'GA001380', 'GA001381', 'GA001382', 'GA001383', 'GA001384', 'GA001385', 'GA001386', 'GA001387', 'GA001388', 'GA001389', 'GA001390', 'GA001391', 'GA001392', 'GA001393', 'GA001394', 'GA001395', 'GA001396', 'GA001397', 'GA001398', 'GA001399', 'GA001400', 'GA001401', 'GA001402', 'GA001403', 'GA001404', 'GA001405', 'GA001406', 'GA001407', 'GA001408', 'GA001409', 'GA001410', 'GA001411', 'GA001412', 'GA001413', 'GA001414', 'GA001415', 'GA001416', 'GA001417', 'GA001418', 'GA001419', 'GA001420', 'GA001421', 'GA001422', 'GA001423', 'GA001424', 'GA001425', 'GA001426', 'GA001427', 'GA001428', 'GA001429', 'GA001430', 'GA001431', 'GA001432', 'GA001433', 'GA001434', 'GA001435', 'GA001436', 'GA001437', 'GA001438', 'GA001439', 'GA001440', 'GA001441', 'GA001442', 'GA001443', 'GA001444', 'GA001445', 'GA001446', 'GA001447', 'GA001448', 'GA001449', 'GA001450', 'GA001451', 'GA001452', 'GA001453', 'GA001454', 'GA001455', 'GA001456', 'GA001457', 'GA001458', 'GA001459', 'GA001460', 'GA001461', 'GA001462', 'GA001463', 'GA001464', 'GA001465', 'GA001466', 'GA001467', 'GA001468', 'GA001469', 'GA001470', 'GA001471', 'GA001472', 'GA001473', 'GA001474', 'GA001475', 'GA001476', 'GA001477', 'GA001478', 'GA001479', 'GA001480', 'GA001481', 'GA001482', 'GA001483', 'GA001484', 'GA001485', 'GA001486', 'GA001487', 'GA001488', 'GA001489', 'GA001490', 'GA001491', 'GA001492', 'GA001493', 'GA001494', 'GA001495', 'GA001496', 'GA001497', 'GA001498', 'GA001499', 'GA001500', 'GA001501', 'GA001502', 'GA001503', 'GA001504', 'GA001505', 'GA001506', 'GA001507', 'GA001508', 'GA001509', 'GA001510', 'GA001511', 'GA001512', 'GA001513', 'GA001514', 'GA001515', 'GA001516', 'GA001517', 'GA001518', 'GA001519', 'GA001520', 'GA001521', 'GA001522', 'GA001523', 'GA001524', 'GA001525', 'GA001526', 'GA001527', 'GA001528', 'GA001529', 'GA001530', 'GA001531', 'GA001532', 'GA001533', 'GA001534', 'GA001535', 'GA001536', 'GA001537', 'GA001538', 'GA001539', 'GA001540', 'GA001541', 'GA001542', 'GA001543', 'GA001544', 'GA001545', 'GA001546', 'GA001547', 'GA001548', 'GA001549', 'GA001550', 'GA001551', 'GA001552', 'GA001553', 'GA001554', 'GA001555', 'GA001556', 'GA001557', 'GA001558', 'GA001559', 'GA001560', 'GA001561', 'GA001562', 'GA001563', 'GA001564', 'GA001565', 'GA001566', 'GA001567', 'GA001568', 'GA001569', 'GA001570', 'GA001571', 'GA001572', 'GA001573', 'GA001574', 'GA001575', 'GA001576', 'GA001577', 'GA001578', 'GA001579', 'GA001580', 'GA001581', 'GA001582', 'GA001583', 'GA001584', 'GA001585', 'GA001586', 'GA001587', 'GA001588', 'GA001589', 'GA001590', 'GA001591', 'GA001592', 'GA001593', 'GA001594', 'GA001595', 'GA001596', 'GA001597', 'GA001598', 'GA001599', 'GA001600', 'GA001601', 'GA001602', 'GA001603', 'GA001604', 'GA001605', 'GA001606', 'GA001607', 'GA001608', 'GA001609', 'GA001610', 'GA001611', 'GA001612', 'GA001613', 'GA001614', 'GA001615', 'GA001616', 'GA001617', 'GA001618', 'GA001619', 'GA001620', 'GA001621', 'GA001622', 'GA001623', 'GA001624', 'GA001625', 'GA001626', 'GA001627', 'GA001628', 'GA001629', 'GA001630', 'GA001631', 'GA001632', 'GA001633', 'GA001634', 'GA001635', 'GA001636', 'GA001637', 'GA001638', 'GA001639', 'GA001640', 'GA001641', 'GA001642', 'GA001643', 'GA001644', 'GA001645', 'GA001646', 'GA001647', 'GA001648', 'GA001649', 'GA001650', 'GA001651', 'GA001652', 'GA001653', 'GA001654', 'GA001655', 'GA001656', 'GA001657', 'GA001658', 'GA001659', 'GA001660', 'GA001661', 'GA001662', 'GA001663', 'GA001664', 'GA001665', 'GA001666', 'GA001667', 'GA001668', 'GA001669', 'GA001670', 'GA001671', 'GA001672', 'GA001673', 'GA001674', 'GA001675', 'GA001676', 'GA001677', 'GA001678', 'GA001679', 'GA001680', 'GA001681', 'GA001682', 'GA001683', 'GA001684', 'GA001685', 'GA001686', 'GA001687', 'GA001688', 'GA001689', 'GA001690', 'GA001691', 'GA001692', 'GA001693', 'GA001694', 'GA001695', 'GA001696', 'GA001697', 'GA001698', 'GA001699', 'GA001700', 'GA001701', 'GA001702', 'GA001703', 'GA001704', 'GA001705', 'GA001706', 'GA001707', 'GA001708', 'GA001709', 'GA001710', 'GA001711', 'GA001712', 'GA001713', 'GA001714', 'GA001715', 'GA001716', 'GA001717', 'GA001718', 'GA001719', 'GA001720', 'GA001721', 'GA001722', 'GA001723', 'GA001724', 'GA001725', 'GA001726', 'GA001727', 'GA001728', 'GA001729', 'GA001730', 'GA001731', 'GA001732', 'GA001733', 'GA001734', 'GA001735', 'GA001736', 'GA001737', 'GA001738', 'GA001739', 'GA001740', 'GA001741', 'GA001742', 'GA001743', 'GA001744', 'GA001745', 'GA001746', 'GA001747', 'GA001748', 'GA001749', 'GA001750', 'GA001751', 'GA001752', 'GA001753', 'GA001754', 'GA001755', 'GA001756', 'GA001757', 'GA001758', 'GA001759', 'GA001760', 'GA001792', 'GA001793', 'GA001794', 'GA001795', 'GA001796', 'GA001797', 'GA001798', 'GA001799', 'GA001800', 'GA001801', 'GA001802', 'GA001803', 'GA001804', 'GA001805', 'GA001806', 'GA001807', 'GA001808', 'GA001809', 'GA001810', 'GA001811', 'GA001812', 'GA001813', 'GA001814', 'GA001815', 'GA001816', 'GA001817', 'GA001818', 'GA001819', 'GA001820', 'GA001821', 'GA001822', 'GA001823', 'GA001824', 'GA001825', 'GA001826', 'GA001827', 'GA001828', 'GA001829', 'GA001830', 'GA001831', 'GA001832', 'GA001834', 'GA001835', 'GA001836', 'GA001837', 'GA001838', 'GA001839', 'GA001841', 'GA001842', 'GA001843', 'GA001844', 'GA001845', 'GA001846', 'GA001847', 'GA001848', 'GA001849', 'GA001850', 'GA001851', 'GA001852', 'GA001853', 'GA001854', 'GA001855', 'GA001856', 'GA001857', 'GA001858', 'GA001859', 'GA001860', 'GA001861', 'GA001862', 'GA001863', 'GA001864', 'GA001865', 'GA001866', 'GA001867', 'GA001869', 'GA001871', 'GA001872', 'GA001873', 'GA001874', 'GA001875', 'GA001876', 'GA001877', 'GA001878', 'GA001880', 'GA001881', 'GA001882', 'GA001883', 'GA001884', 'GA001885', 'GA001886', 'GA001887', 'GA001888', 'GA001890', 'GA001891', 'GA001892', 'GA001893', 'GA001894', 'GA001895']\n"
     ]
    }
   ],
   "execution_count": 9
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-18T17:16:54.915460Z",
     "start_time": "2025-05-18T17:16:53.988706Z"
    }
   },
   "cell_type": "code",
   "source": [
    "df = pd.read_csv(vcf_file3, sep='\\t', comment='#', names=fixed_columns)\n",
    "#if csv file uncomment this and comment above\n",
    "#df = pd.read_csv(vcf_file3, sep=\",\")\n",
    "df.head()  # Column index 1 = POS, use POS as we know the cols\n"
   ],
   "id": "a938839036e2bb15",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "  CHROM      POS         ID REF  ALT     QUAL FILTER  \\\n",
       "0     Y  2649666          .   A  C,G   362.13   PASS   \n",
       "1     Y  2658271  rs1800865   G  A,*  1242.49   PASS   \n",
       "2     Y  2667427          .   G  T,A  3869.90   PASS   \n",
       "3     Y  2688624          .   T  *,A  6422.56   PASS   \n",
       "4     Y  2688626          .   T  *,A  6294.31   PASS   \n",
       "\n",
       "                          INFO          FORMAT                GA000431  ...  \\\n",
       "0           GT1168=0:621,.:542  GT:AD:DP:GQ:PL  0:3,0,0:3:99:0,125,125  ...   \n",
       "1           GT1168=0:630,.:533  GT:AD:DP:GQ:PL  0:9,0,0:9:99:0,246,246  ...   \n",
       "2           GT1168=0:630,.:533  GT:AD:DP:GQ:PL  0:7,0,0:7:99:0,166,166  ...   \n",
       "3  GT1168=2:6,1:52,0:218,.:887  GT:AD:DP:GQ:PL           .:0,0,0:.:.:.  ...   \n",
       "4  GT1168=2:6,1:51,0:272,.:834  GT:AD:DP:GQ:PL           .:0,0,0:.:.:.  ...   \n",
       "\n",
       "                 GA001885                GA001886                GA001887  \\\n",
       "0  0:5,0,0:5:99:0,117,117  0:4,0,0:4:99:0,135,135  0:3,0,0:3:99:0,108,108   \n",
       "1  0:3,0,0:3:99:0,128,128  0:4,0,0:4:99:0,109,109  0:5,0,0:5:99:0,108,108   \n",
       "2  0:5,0,0:5:99:0,180,180  0:4,0,0:4:99:0,135,135  0:4,0,0:4:99:0,135,135   \n",
       "3           .:0,0,0:.:.:.    0:7,0,0:7:45:0,45,45           .:0,0,0:.:.:.   \n",
       "4           .:0,0,0:.:.:.    0:7,0,0:7:45:0,45,45           .:0,0,0:.:.:.   \n",
       "\n",
       "                 GA001888       GA001890                GA001891  \\\n",
       "0    0:3,0,0:3:99:0,99,99  .:0,0,0:.:.:.  0:3,0,0:3:99:0,108,108   \n",
       "1  0:5,0,0:5:99:0,104,104  .:0,0,0:.:.:.  0:4,0,0:4:99:0,104,104   \n",
       "2  0:4,0,0:4:99:0,128,128  .:0,0,0:.:.:.  0:4,0,0:4:99:0,102,102   \n",
       "3    0:2,0,0:2:83:0,83,83  .:0,0,0:.:.:.           .:0,0,0:.:.:.   \n",
       "4    0:2,0,0:2:83:0,83,83  .:0,0,0:.:.:.           .:0,0,0:.:.:.   \n",
       "\n",
       "                 GA001892                GA001893                GA001894  \\\n",
       "0    0:2,0,0:2:83:0,83,83  0:3,0,0:3:99:0,104,104    0:2,0,0:2:88:0,88,88   \n",
       "1  0:4,0,0:4:99:0,123,123  0:5,0,0:5:99:0,135,135  0:4,0,0:4:99:0,128,128   \n",
       "2  0:7,0,0:7:99:0,225,225  0:4,0,0:4:99:0,130,130  0:5,0,0:5:99:0,108,108   \n",
       "3           .:0,0,0:.:.:.           .:0,0,0:.:.:.           .:0,0,0:.:.:.   \n",
       "4           .:0,0,0:.:.:.           .:0,0,0:.:.:.           .:0,0,0:.:.:.   \n",
       "\n",
       "        GA001895  \n",
       "0  .:0,0,0:.:.:.  \n",
       "1  .:0,0,0:.:.:.  \n",
       "2  .:0,0,0:.:.:.  \n",
       "3  .:0,0,0:.:.:.  \n",
       "4  .:0,0,0:.:.:.  \n",
       "\n",
       "[5 rows x 1172 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>GA000431</th>\n",
       "      <th>...</th>\n",
       "      <th>GA001885</th>\n",
       "      <th>GA001886</th>\n",
       "      <th>GA001887</th>\n",
       "      <th>GA001888</th>\n",
       "      <th>GA001890</th>\n",
       "      <th>GA001891</th>\n",
       "      <th>GA001892</th>\n",
       "      <th>GA001893</th>\n",
       "      <th>GA001894</th>\n",
       "      <th>GA001895</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Y</td>\n",
       "      <td>2649666</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>C,G</td>\n",
       "      <td>362.13</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT1168=0:621,.:542</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0:3,0,0:3:99:0,125,125</td>\n",
       "      <td>...</td>\n",
       "      <td>0:5,0,0:5:99:0,117,117</td>\n",
       "      <td>0:4,0,0:4:99:0,135,135</td>\n",
       "      <td>0:3,0,0:3:99:0,108,108</td>\n",
       "      <td>0:3,0,0:3:99:0,99,99</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "      <td>0:3,0,0:3:99:0,108,108</td>\n",
       "      <td>0:2,0,0:2:83:0,83,83</td>\n",
       "      <td>0:3,0,0:3:99:0,104,104</td>\n",
       "      <td>0:2,0,0:2:88:0,88,88</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Y</td>\n",
       "      <td>2658271</td>\n",
       "      <td>rs1800865</td>\n",
       "      <td>G</td>\n",
       "      <td>A,*</td>\n",
       "      <td>1242.49</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT1168=0:630,.:533</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0:9,0,0:9:99:0,246,246</td>\n",
       "      <td>...</td>\n",
       "      <td>0:3,0,0:3:99:0,128,128</td>\n",
       "      <td>0:4,0,0:4:99:0,109,109</td>\n",
       "      <td>0:5,0,0:5:99:0,108,108</td>\n",
       "      <td>0:5,0,0:5:99:0,104,104</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "      <td>0:4,0,0:4:99:0,104,104</td>\n",
       "      <td>0:4,0,0:4:99:0,123,123</td>\n",
       "      <td>0:5,0,0:5:99:0,135,135</td>\n",
       "      <td>0:4,0,0:4:99:0,128,128</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Y</td>\n",
       "      <td>2667427</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>T,A</td>\n",
       "      <td>3869.90</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT1168=0:630,.:533</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>0:7,0,0:7:99:0,166,166</td>\n",
       "      <td>...</td>\n",
       "      <td>0:5,0,0:5:99:0,180,180</td>\n",
       "      <td>0:4,0,0:4:99:0,135,135</td>\n",
       "      <td>0:4,0,0:4:99:0,135,135</td>\n",
       "      <td>0:4,0,0:4:99:0,128,128</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "      <td>0:4,0,0:4:99:0,102,102</td>\n",
       "      <td>0:7,0,0:7:99:0,225,225</td>\n",
       "      <td>0:4,0,0:4:99:0,130,130</td>\n",
       "      <td>0:5,0,0:5:99:0,108,108</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Y</td>\n",
       "      <td>2688624</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>*,A</td>\n",
       "      <td>6422.56</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT1168=2:6,1:52,0:218,.:887</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "      <td>...</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "      <td>0:7,0,0:7:45:0,45,45</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "      <td>0:2,0,0:2:83:0,83,83</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Y</td>\n",
       "      <td>2688626</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>*,A</td>\n",
       "      <td>6294.31</td>\n",
       "      <td>PASS</td>\n",
       "      <td>GT1168=2:6,1:51,0:272,.:834</td>\n",
       "      <td>GT:AD:DP:GQ:PL</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "      <td>...</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "      <td>0:7,0,0:7:45:0,45,45</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "      <td>0:2,0,0:2:83:0,83,83</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "      <td>.:0,0,0:.:.:.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1172 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 10
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "## Applying SNP conditions\n",
    "- QUAL > 30\n",
    "- MAF > 0.05\n",
    "- INFO > 10\n",
    "- Single nucleotide in REF, ALT"
   ],
   "id": "e1f4458ab6a8dc75"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Set the conditions\n",
    "QUALITY = df[\"QUAL\"].astype('int') >= 30\n",
    "\n",
    "## Take out the numbers from DP=xx;AP=yy using generator expression\n",
    "if df['INFO'].dtype == 'object':\n",
    "    #df['AF'] = [float(s.split(';')[1].split('=')[1]) for s in np.array(df[\"INFO\"].tolist())]\n",
    "    df['INFO'] = [int(s.split(';')[0].split('=')[1]) for s in np.array(df[\"INFO\"].tolist())]\n",
    "\n",
    "INFO = df['INFO'] > 10\n",
    "#AF = df['AF'] > 0.05\n",
    "\n",
    "filt = QUALITY & INFO #& AF\n",
    "\n",
    "## Check REF and ALT\n",
    "\n",
    "\n",
    "df.loc[filt, ['CHROM', 'POS', 'REF', 'ALT', 'QUAL', 'INFO', 'FILTER']]"
   ],
   "id": "1f5dd83ad5eb0bd4",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-04T15:22:52.649822Z",
     "start_time": "2025-05-04T15:22:52.638413Z"
    }
   },
   "cell_type": "code",
   "source": [
    "     # Save the file test_Y_chrom_snp_positions\n",
    "df = df.loc[filt, ['CHROM', 'POS', 'REF', 'ALT', 'QUAL', 'INFO', 'FILTER']]\n",
    "df.to_csv(\"test_1_chrom_snp_positions.csv\", index=False)\n",
    "print(f\"Saved successfully! with shape {df.shape}\")"
   ],
   "id": "c3593c58a1ca3016",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved successfully! with shape (7, 7)\n"
     ]
    }
   ],
   "execution_count": 104
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-06T09:01:52.504350Z",
     "start_time": "2025-05-06T09:01:52.083748Z"
    }
   },
   "cell_type": "code",
   "source": "import pandas as pd",
   "id": "6ad3998949931e20",
   "outputs": [],
   "execution_count": 9
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Import Pandas",
   "id": "d8ca47acaa68ca24"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-06T09:01:54.034635Z",
     "start_time": "2025-05-06T09:01:54.027098Z"
    }
   },
   "cell_type": "code",
   "source": [
    "vcf_file = \"homo_sapiens-chrY.vcf.gz\"\n",
    "vcf_file1 = \"GAsP_1163Samples.InDels.Feb2020.vcf.gz\"\n",
    "vcf_file2 = r\"I:\\Genomic Project\\GAsP_1163Samples.InDels.Feb2020.vcf.gz\"\n"
   ],
   "id": "90a592e920c934e3",
   "outputs": [],
   "execution_count": 10
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Homo_sapiens-chrY VCF file GRC38",
   "id": "3f778a7237ca939f"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "790cceb1ca8409bb"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-06T09:01:56.101192Z",
     "start_time": "2025-05-06T09:01:56.088376Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def get_vcf_header(vcf_file_input):\n",
    "    import gzip\n",
    "    header_line = \"\"\n",
    "    # set the opener if it is gzip\n",
    "    opener = gzip.open if vcf_file_input.endswith(\".gz\") else open\n",
    "    # open the file and find the header line\n",
    "    with opener(vcf_file_input, \"rt\") as file:\n",
    "        for line in file:\n",
    "            if line.startswith(\"#CHROM\"):\n",
    "                header_line = line\n",
    "                break\n",
    "            elif not line.startswith(\"#\"):\n",
    "                break\n",
    "    # format the header line to a list format\n",
    "    if header_line:\n",
    "        header_columns = header_line.strip(\"#\").strip(\"\\n\").split(\"\\t\")\n",
    "        return header_columns\n",
    "    else:\n",
    "        return \"Unable to find the header\"\n",
    "\n",
    "fixed_columns = get_vcf_header(vcf_file)  # change file name here >>\n",
    "print(f\"These are the columns found :\\n{fixed_columns}\")\n"
   ],
   "id": "4d9cbce58527369a",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "These are the columns found :\n",
      "['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO']\n"
     ]
    }
   ],
   "execution_count": 11
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Python Codes for Column Header",
   "id": "387a6605778210e4"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-06T09:02:01.317519Z",
     "start_time": "2025-05-06T09:01:59.415486Z"
    }
   },
   "cell_type": "code",
   "source": [
    "df = pd.read_csv(vcf_file, sep='\\t', comment='#', names=fixed_columns)\n",
    "df.head()  # Column index 1 = POS, use POS as we know the cols\n",
    "\n"
   ],
   "id": "f773ddc450515c0b",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "  CHROM    POS            ID REF  ALT QUAL FILTER  \\\n",
       "0     Y  10002  rs1226858834   T    A    .      .   \n",
       "1     Y  10003   rs375039031   A  C,G    .      .   \n",
       "2     Y  10007  rs1422184628   C    G    .      .   \n",
       "3     Y  10008  rs1179917603   T    G    .      .   \n",
       "4     Y  10009   rs565284081   A  C,G    .      .   \n",
       "\n",
       "                                INFO  \n",
       "0           dbSNP_156;TSA=SNV;E_Freq  \n",
       "1  dbSNP_156;TSA=SNV;E_Freq;E_gnomAD  \n",
       "2           dbSNP_156;TSA=SNV;E_Freq  \n",
       "3  dbSNP_156;TSA=SNV;E_Freq;E_gnomAD  \n",
       "4  dbSNP_156;TSA=SNV;E_Freq;E_gnomAD  "
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Y</td>\n",
       "      <td>10002</td>\n",
       "      <td>rs1226858834</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Y</td>\n",
       "      <td>10003</td>\n",
       "      <td>rs375039031</td>\n",
       "      <td>A</td>\n",
       "      <td>C,G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq;E_gnomAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Y</td>\n",
       "      <td>10007</td>\n",
       "      <td>rs1422184628</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Y</td>\n",
       "      <td>10008</td>\n",
       "      <td>rs1179917603</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq;E_gnomAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Y</td>\n",
       "      <td>10009</td>\n",
       "      <td>rs565284081</td>\n",
       "      <td>A</td>\n",
       "      <td>C,G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq;E_gnomAD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 12
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Y Chromosome File",
   "id": "bc48801c17587532"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-06T09:02:04.796178Z",
     "start_time": "2025-05-06T09:02:04.775951Z"
    }
   },
   "cell_type": "code",
   "source": [
    "df_snp = df['POS']\n",
    "print(\"SNP values: \")\n",
    "df_snp.head()\n"
   ],
   "id": "eb07b598d344a0dd",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNP values: \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0    10002\n",
       "1    10003\n",
       "2    10007\n",
       "3    10008\n",
       "4    10009\n",
       "Name: POS, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 13
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# SNP Positions from Individual Y chromosome file",
   "id": "91c4e923101dd834"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-06T09:02:28.359163Z",
     "start_time": "2025-05-06T09:02:27.592227Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Save the SNPs to a file\n",
    "df_snp.to_csv(\"snp_positions.csv\", index=False, header=False)\n"
   ],
   "id": "f0d6ec9527fa3396",
   "outputs": [],
   "execution_count": 16
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-06T09:02:30.588267Z",
     "start_time": "2025-05-06T09:02:30.581282Z"
    }
   },
   "cell_type": "code",
   "source": "df_snp = df_snp.head(500)",
   "id": "4ed5ed36ad2b395b",
   "outputs": [],
   "execution_count": 17
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# finding SNPs clusters",
   "id": "98093f1ab55632d8"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-06T09:02:34.100470Z",
     "start_time": "2025-05-06T09:02:33.982991Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Convert the POS col to a list\n",
    "snp_pos_list = df['POS'].sort_values().reset_index(drop=True).tolist()\n"
   ],
   "id": "aa72a136160c876e",
   "outputs": [],
   "execution_count": 18
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-06T09:04:53.337658Z",
     "start_time": "2025-05-06T09:04:53.324607Z"
    }
   },
   "cell_type": "code",
   "source": [
    "MAX_DISTANCE = 200\n",
    "SNP_NUM = 3\n",
    "snp_cluster_list = []\n",
    "temp_cluster_list = [snp_pos_list[0]]\n",
    "\n",
    "for i in range(1, len(snp_pos_list[0:50])):\n",
    "    if (snp_pos_list[i] - snp_pos_list[i-1]) <= MAX_DISTANCE:\n",
    "        temp_cluster_list.append(snp_pos_list[i])\n",
    "    else:\n",
    "        if len(temp_cluster_list) >= SNP_NUM:\n",
    "            snp_cluster_list.append(temp_cluster_list)\n",
    "            # update the temp_cluster_list with the recent element\n",
    "            temp_cluster_list = [snp_pos_list[i]]\n",
    "\n",
    "# If at the end of the loop the size of temp_cluster_list is more than 3\n",
    "if len(temp_cluster_list) >= SNP_NUM:\n",
    "    snp_cluster_list.append(temp_cluster_list)\n",
    "\n",
    "print(f\"Cluster numbers: {snp_cluster_list}\")\n"
   ],
   "id": "9436b662976f26b4",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster numbers: [[10002, 10003, 10007, 10008, 10009, 10010, 10014, 10015, 10015, 10016, 10017, 10020, 10021, 10021, 10021, 10022, 10023, 10023, 10023, 10025, 10026, 10027, 10028, 10029, 10029, 10029, 10030, 10031, 10032, 10033, 10034, 10035, 10035, 10035, 10036, 10036, 10037, 10037, 10038, 10039, 10042, 10046, 10047, 10048, 10049, 10050, 10051, 10052, 10053, 10055]]\n"
     ]
    }
   ],
   "execution_count": 21
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "",
   "id": "972ca10b5949fc96"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Identify Microhaplotypes Positions",
   "id": "d7a1f87076278e44"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-05-06T09:04:57.832622Z",
     "start_time": "2025-05-06T09:04:57.788181Z"
    }
   },
   "cell_type": "code",
   "source": [
    "filt = df['POS'].isin(snp_cluster_list[0])\n",
    "df[filt]\n"
   ],
   "id": "541d8bfa3e368153",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   CHROM    POS            ID         REF          ALT QUAL FILTER  \\\n",
       "0      Y  10002  rs1226858834           T            A    .      .   \n",
       "1      Y  10003   rs375039031           A          C,G    .      .   \n",
       "2      Y  10007  rs1422184628           C            G    .      .   \n",
       "3      Y  10008  rs1179917603           T            G    .      .   \n",
       "4      Y  10009   rs565284081           A          C,G    .      .   \n",
       "5      Y  10010   rs867361848           A            G    .      .   \n",
       "6      Y  10014   rs370048753           T          A,G    .      .   \n",
       "7      Y  10015   rs113469508           A          C,G    .      .   \n",
       "8      Y  10015  rs2068137778           A           AA    .      .   \n",
       "9      Y  10016  rs1423076556           A            G    .      .   \n",
       "10     Y  10017  rs1232177317        CCCT            C    .      .   \n",
       "11     Y  10020  rs1307895415           T          C,G    .      .   \n",
       "12     Y  10021  rs1557356458           A         AACA    .      .   \n",
       "13     Y  10021   rs376157089           A          C,G    .      .   \n",
       "14     Y  10021   rs745593600           A           AA    .      .   \n",
       "15     Y  10022  rs1331379320           A          C,G    .      .   \n",
       "16     Y  10023  rs2068137914           C           CC    .      .   \n",
       "17     Y  10023  rs2068137924           C           CA    .      .   \n",
       "18     Y  10023   rs867417461           C          A,T    .      .   \n",
       "19     Y  10025   rs866208202           C            T    .      .   \n",
       "20     Y  10026   rs774535218           T            C    .      .   \n",
       "21     Y  10027   rs866812978           A          C,G    .      .   \n",
       "22     Y  10028   rs868775747           A          C,G    .      .   \n",
       "23     Y  10029  rs1168920679        CCCT            C    .      .   \n",
       "24     Y  10029  rs1462779560           C           CC    .      .   \n",
       "25     Y  10029   rs866418186           C            A    .      .   \n",
       "26     Y  10030  rs1477801263           C          A,G    .      .   \n",
       "27     Y  10031  rs1190220124           C            T    .      .   \n",
       "28     Y  10032   rs748268487           T          A,C    .      .   \n",
       "29     Y  10033  rs1233577647           A        C,G,T    .      .   \n",
       "30     Y  10034   rs987352364           A          C,G    .      .   \n",
       "31     Y  10035   rs201278642         CCC         C,CC    .      .   \n",
       "32     Y  10035  rs2068138143           C            A    .      .   \n",
       "33     Y  10035  rs2068138210           C  CA,CAA,CTAA    .      .   \n",
       "34     Y  10036  rs1234515714           C            A    .      .   \n",
       "35     Y  10036  rs1293842805        CCTC            C    .      .   \n",
       "36     Y  10037  rs1557356473           C        CTAAC    .      .   \n",
       "37     Y  10037  rs2068138224           C            T    .      .   \n",
       "38     Y  10038  rs2068138306  TCTGAAAGTG            T    .      .   \n",
       "39     Y  10039  rs1277006941           C          A,G    .      .   \n",
       "40     Y  10042  rs2068138317           A            T    .      .   \n",
       "41     Y  10046  rs2068138334           T            C    .      .   \n",
       "42     Y  10047  rs2068138346           G            C    .      .   \n",
       "43     Y  10048  rs1311505857           G          A,C    .      .   \n",
       "44     Y  10049   rs973139367           A            T    .      .   \n",
       "45     Y  10050   rs920054068           C            G    .      .   \n",
       "46     Y  10051  rs2068138417           C            T    .      .   \n",
       "47     Y  10052   rs186434315           T          A,C    .      .   \n",
       "48     Y  10053   rs931368434           A          G,T    .      .   \n",
       "49     Y  10055  rs2068138471           C            G    .      .   \n",
       "\n",
       "                                         INFO  \n",
       "0                    dbSNP_156;TSA=SNV;E_Freq  \n",
       "1           dbSNP_156;TSA=SNV;E_Freq;E_gnomAD  \n",
       "2                    dbSNP_156;TSA=SNV;E_Freq  \n",
       "3           dbSNP_156;TSA=SNV;E_Freq;E_gnomAD  \n",
       "4           dbSNP_156;TSA=SNV;E_Freq;E_gnomAD  \n",
       "5                    dbSNP_156;TSA=SNV;E_Freq  \n",
       "6           dbSNP_156;TSA=SNV;E_Freq;E_gnomAD  \n",
       "7           dbSNP_156;TSA=SNV;E_Freq;E_gnomAD  \n",
       "8                  dbSNP_156;TSA=indel;E_Freq  \n",
       "9           dbSNP_156;TSA=SNV;E_Freq;E_gnomAD  \n",
       "10              dbSNP_156;TSA=deletion;E_Freq  \n",
       "11                   dbSNP_156;TSA=SNV;E_Freq  \n",
       "12         dbSNP_156;TSA=indel;E_Freq;E_1000G  \n",
       "13          dbSNP_156;TSA=SNV;E_Freq;E_gnomAD  \n",
       "14         dbSNP_156;TSA=indel;E_Freq;E_1000G  \n",
       "15          dbSNP_156;TSA=SNV;E_Freq;E_gnomAD  \n",
       "16                 dbSNP_156;TSA=indel;E_Freq  \n",
       "17             dbSNP_156;TSA=insertion;E_Freq  \n",
       "18                   dbSNP_156;TSA=SNV;E_Freq  \n",
       "19                   dbSNP_156;TSA=SNV;E_Freq  \n",
       "20  dbSNP_156;TSA=SNV;E_Freq;E_1000G;E_gnomAD  \n",
       "21          dbSNP_156;TSA=SNV;E_Freq;E_gnomAD  \n",
       "22          dbSNP_156;TSA=SNV;E_Freq;E_gnomAD  \n",
       "23              dbSNP_156;TSA=deletion;E_Freq  \n",
       "24                 dbSNP_156;TSA=indel;E_Freq  \n",
       "25                   dbSNP_156;TSA=SNV;E_Freq  \n",
       "26                   dbSNP_156;TSA=SNV;E_Freq  \n",
       "27                   dbSNP_156;TSA=SNV;E_Freq  \n",
       "28  dbSNP_156;TSA=SNV;E_Freq;E_1000G;E_gnomAD  \n",
       "29          dbSNP_156;TSA=SNV;E_Freq;E_gnomAD  \n",
       "30                   dbSNP_156;TSA=SNV;E_Freq  \n",
       "31         dbSNP_156;TSA=indel;E_Freq;E_1000G  \n",
       "32                   dbSNP_156;TSA=SNV;E_Freq  \n",
       "33             dbSNP_156;TSA=insertion;E_Freq  \n",
       "34                   dbSNP_156;TSA=SNV;E_Freq  \n",
       "35                 dbSNP_156;TSA=indel;E_Freq  \n",
       "36                 dbSNP_156;TSA=indel;E_Freq  \n",
       "37                   dbSNP_156;TSA=SNV;E_Freq  \n",
       "38              dbSNP_156;TSA=deletion;E_Freq  \n",
       "39          dbSNP_156;TSA=SNV;E_Freq;E_gnomAD  \n",
       "40                   dbSNP_156;TSA=SNV;E_Freq  \n",
       "41                   dbSNP_156;TSA=SNV;E_Freq  \n",
       "42                   dbSNP_156;TSA=SNV;E_Freq  \n",
       "43          dbSNP_156;TSA=SNV;E_Freq;E_TOPMed  \n",
       "44                   dbSNP_156;TSA=SNV;E_Freq  \n",
       "45                   dbSNP_156;TSA=SNV;E_Freq  \n",
       "46                   dbSNP_156;TSA=SNV;E_Freq  \n",
       "47  dbSNP_156;TSA=SNV;E_Freq;E_1000G;E_gnomAD  \n",
       "48                   dbSNP_156;TSA=SNV;E_Freq  \n",
       "49                   dbSNP_156;TSA=SNV;E_Freq  "
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Y</td>\n",
       "      <td>10002</td>\n",
       "      <td>rs1226858834</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Y</td>\n",
       "      <td>10003</td>\n",
       "      <td>rs375039031</td>\n",
       "      <td>A</td>\n",
       "      <td>C,G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq;E_gnomAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Y</td>\n",
       "      <td>10007</td>\n",
       "      <td>rs1422184628</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Y</td>\n",
       "      <td>10008</td>\n",
       "      <td>rs1179917603</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq;E_gnomAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Y</td>\n",
       "      <td>10009</td>\n",
       "      <td>rs565284081</td>\n",
       "      <td>A</td>\n",
       "      <td>C,G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq;E_gnomAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Y</td>\n",
       "      <td>10010</td>\n",
       "      <td>rs867361848</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Y</td>\n",
       "      <td>10014</td>\n",
       "      <td>rs370048753</td>\n",
       "      <td>T</td>\n",
       "      <td>A,G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq;E_gnomAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Y</td>\n",
       "      <td>10015</td>\n",
       "      <td>rs113469508</td>\n",
       "      <td>A</td>\n",
       "      <td>C,G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq;E_gnomAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Y</td>\n",
       "      <td>10015</td>\n",
       "      <td>rs2068137778</td>\n",
       "      <td>A</td>\n",
       "      <td>AA</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=indel;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Y</td>\n",
       "      <td>10016</td>\n",
       "      <td>rs1423076556</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq;E_gnomAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Y</td>\n",
       "      <td>10017</td>\n",
       "      <td>rs1232177317</td>\n",
       "      <td>CCCT</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=deletion;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Y</td>\n",
       "      <td>10020</td>\n",
       "      <td>rs1307895415</td>\n",
       "      <td>T</td>\n",
       "      <td>C,G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Y</td>\n",
       "      <td>10021</td>\n",
       "      <td>rs1557356458</td>\n",
       "      <td>A</td>\n",
       "      <td>AACA</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=indel;E_Freq;E_1000G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Y</td>\n",
       "      <td>10021</td>\n",
       "      <td>rs376157089</td>\n",
       "      <td>A</td>\n",
       "      <td>C,G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq;E_gnomAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Y</td>\n",
       "      <td>10021</td>\n",
       "      <td>rs745593600</td>\n",
       "      <td>A</td>\n",
       "      <td>AA</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=indel;E_Freq;E_1000G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Y</td>\n",
       "      <td>10022</td>\n",
       "      <td>rs1331379320</td>\n",
       "      <td>A</td>\n",
       "      <td>C,G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq;E_gnomAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Y</td>\n",
       "      <td>10023</td>\n",
       "      <td>rs2068137914</td>\n",
       "      <td>C</td>\n",
       "      <td>CC</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=indel;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Y</td>\n",
       "      <td>10023</td>\n",
       "      <td>rs2068137924</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=insertion;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Y</td>\n",
       "      <td>10023</td>\n",
       "      <td>rs867417461</td>\n",
       "      <td>C</td>\n",
       "      <td>A,T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Y</td>\n",
       "      <td>10025</td>\n",
       "      <td>rs866208202</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Y</td>\n",
       "      <td>10026</td>\n",
       "      <td>rs774535218</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq;E_1000G;E_gnomAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Y</td>\n",
       "      <td>10027</td>\n",
       "      <td>rs866812978</td>\n",
       "      <td>A</td>\n",
       "      <td>C,G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq;E_gnomAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Y</td>\n",
       "      <td>10028</td>\n",
       "      <td>rs868775747</td>\n",
       "      <td>A</td>\n",
       "      <td>C,G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq;E_gnomAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Y</td>\n",
       "      <td>10029</td>\n",
       "      <td>rs1168920679</td>\n",
       "      <td>CCCT</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=deletion;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Y</td>\n",
       "      <td>10029</td>\n",
       "      <td>rs1462779560</td>\n",
       "      <td>C</td>\n",
       "      <td>CC</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=indel;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Y</td>\n",
       "      <td>10029</td>\n",
       "      <td>rs866418186</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Y</td>\n",
       "      <td>10030</td>\n",
       "      <td>rs1477801263</td>\n",
       "      <td>C</td>\n",
       "      <td>A,G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Y</td>\n",
       "      <td>10031</td>\n",
       "      <td>rs1190220124</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Y</td>\n",
       "      <td>10032</td>\n",
       "      <td>rs748268487</td>\n",
       "      <td>T</td>\n",
       "      <td>A,C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq;E_1000G;E_gnomAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Y</td>\n",
       "      <td>10033</td>\n",
       "      <td>rs1233577647</td>\n",
       "      <td>A</td>\n",
       "      <td>C,G,T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq;E_gnomAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Y</td>\n",
       "      <td>10034</td>\n",
       "      <td>rs987352364</td>\n",
       "      <td>A</td>\n",
       "      <td>C,G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Y</td>\n",
       "      <td>10035</td>\n",
       "      <td>rs201278642</td>\n",
       "      <td>CCC</td>\n",
       "      <td>C,CC</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=indel;E_Freq;E_1000G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Y</td>\n",
       "      <td>10035</td>\n",
       "      <td>rs2068138143</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Y</td>\n",
       "      <td>10035</td>\n",
       "      <td>rs2068138210</td>\n",
       "      <td>C</td>\n",
       "      <td>CA,CAA,CTAA</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=insertion;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Y</td>\n",
       "      <td>10036</td>\n",
       "      <td>rs1234515714</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Y</td>\n",
       "      <td>10036</td>\n",
       "      <td>rs1293842805</td>\n",
       "      <td>CCTC</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=indel;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Y</td>\n",
       "      <td>10037</td>\n",
       "      <td>rs1557356473</td>\n",
       "      <td>C</td>\n",
       "      <td>CTAAC</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=indel;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Y</td>\n",
       "      <td>10037</td>\n",
       "      <td>rs2068138224</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Y</td>\n",
       "      <td>10038</td>\n",
       "      <td>rs2068138306</td>\n",
       "      <td>TCTGAAAGTG</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=deletion;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Y</td>\n",
       "      <td>10039</td>\n",
       "      <td>rs1277006941</td>\n",
       "      <td>C</td>\n",
       "      <td>A,G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq;E_gnomAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Y</td>\n",
       "      <td>10042</td>\n",
       "      <td>rs2068138317</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Y</td>\n",
       "      <td>10046</td>\n",
       "      <td>rs2068138334</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Y</td>\n",
       "      <td>10047</td>\n",
       "      <td>rs2068138346</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Y</td>\n",
       "      <td>10048</td>\n",
       "      <td>rs1311505857</td>\n",
       "      <td>G</td>\n",
       "      <td>A,C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq;E_TOPMed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Y</td>\n",
       "      <td>10049</td>\n",
       "      <td>rs973139367</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Y</td>\n",
       "      <td>10050</td>\n",
       "      <td>rs920054068</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Y</td>\n",
       "      <td>10051</td>\n",
       "      <td>rs2068138417</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Y</td>\n",
       "      <td>10052</td>\n",
       "      <td>rs186434315</td>\n",
       "      <td>T</td>\n",
       "      <td>A,C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq;E_1000G;E_gnomAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Y</td>\n",
       "      <td>10053</td>\n",
       "      <td>rs931368434</td>\n",
       "      <td>A</td>\n",
       "      <td>G,T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Y</td>\n",
       "      <td>10055</td>\n",
       "      <td>rs2068138471</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>dbSNP_156;TSA=SNV;E_Freq</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 22
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Display 49 Microhaplotypes",
   "id": "ca2de86ac21e2590"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
